scripts/source-maintenance.sh

```diff
@@ -66,7 +66,8 @@ for FILENAME in `git ls-files`; do
 	#
 	# Code Style formatting maintenance
 	#
-        if test "${ASVER}"; then
+	awk -f ./scripts/maintenance.awk "${FILENAME}" >"${FILENAME}.awkbak" && mv "${FILENAME}.awkbak" "${FILENAME}"
```

Comment:
We should iterate through individual scripts named using a hard-coded pattern (e.g. `./scripts/maintenance-cpp-*` or `./scripts/maintenance/cpp-*`) rather than dumping all adjustments into one script.

The script execution command should be made interpreter-neutral (rather than forcing AWK usage for every adjustment script). The script itself should supply its interpreter name (and invocation parameters).

Finally, existing script(s) should be adjusted to use the new approach.

Reply:
How do you feel about adding dozens upon dozens of files containing one line of code each and which developers are then asked to run manually over their code? Sounds like a lot of annoyance for the sake of having this one PR meet your personal definition of perfection.

The reason for choosing AWK here over things like perl is that the rules are essentially just a regex pattern with some action associated.

Also, note that each of these files needs the shell command to parse and process the entire source code input file. Having two scripts would mean twice the processing for each file in the repository, three scripts thrice the processing and so on increasing for the number of individual changes being performed _ever_.

Reply:
> How do you feel about adding dozens upon dozens of files containing one line of code each and which developers are then asked to run manually over their code?

Feels like a straw man argument: Nobody suggested adding dozens of files or manual execution.

> The reason for choosing AWK here over things like perl is that the rules are essentially just a regex pattern with some action associated.

We should not be discussing advantages of one language over the other until we have the rules  that the scripts will implement, especially when each script can select its own "best" interpreter.

> Also, note that each of these files needs the shell command to parse and process the entire source code input file. Having two scripts would mean twice the processing for each file in the repository ...

Yes, the processing speed will decrease with the number of scripts, but we already execute a few scripts per file, and I do not expect us to add a lot more scripts.

Reply:
> > How do you feel about adding dozens upon dozens of files containing one line of code each and which developers are then asked to run manually over their code?
> 
> Feels like a straw man argument: Nobody suggested adding dozens of files or manual execution.
> 

You suggested breaking this one script down into individual scripts applied using a pattern. That mess is the natural mid-to-long term outcome if each maintenance change needs its own script.

> > The reason for choosing AWK here over things like perl is that the rules are essentially just a regex pattern with some action associated.
> 
> We should not be discussing advantages of one language over the other until we have the rules that the scripts will implement, especially when each script can select its own "best" interpreter.
> 

Fair point.

> > Also, note that each of these files needs the shell command to parse and process the entire source code input file. Having two scripts would mean twice the processing for each file in the repository ...
> 
> Yes, the processing speed will decrease with the number of scripts, but we already execute a few scripts per file, and I do not expect us to add a lot more scripts.

This counter argument depends on that "feeling" about there never being many scripts. I expect there to be a great many scripts added if we split these up into multiple. With additional work some of the existing scripts may actually be removable, but the future will see about that.

Reply:
> You suggested breaking this one script down into individual scripts applied using a pattern.

I did not.

Reply:
> > You suggested breaking this one script down into individual scripts applied using a pattern.
> 
> I did not.

The first sentence of your review comment was:
> We should iterate through individual scripts named using a hard-coded pattern ... rather than dumping all adjustments into one script.

Does that not mean placing the planned adjustments from this script into individual files and iterating through those?

Reply:
This continuing misunderstanding is driven, in part, by the lack of (shared) specifics regarding the future application of the proposed source code adjustment script(s). At first, there were no specifics at all. Then you mentioned HERE/NULL removal but did not like me saying that this is what the script(s) will do -- it is just an example of what it could do. So the future remains fuzzy, but here is my understanding of where we are going:

Squid will continue to make a few source code adjustments. Those adjustments can be naturally split into several scripts. We have at least one such script today. Those scripts will all be executed against each git-controlled C++-related file (for now). The names of those scripts will be hard-coded (as in the current PR) or identified using a simple filename pattern (as I have suggested). Please note that my primary concerns are not about that difference.

Reply:
Very well. The proposed long-term contents of this script would be rules along the lines of these:

```
# FooDisplay macro is deprecated
/FooDisplay/ { gsub("FooDisplay","foo->print"); }

# BarStateData got renamed to Foo::Bar
/BarStateData/ { gsub("BarClass","Foo::Bar"); }

# NULL is deprectated by nullptr in C++11 code
# auto-remove common cases
/NULL/ {
  if ( FILENAME ~ /(src\/|tools\/|\.cci?)$/ ) {
    gsub(/[ ]=[ ]?NULL[;]/, " = nullptr;");
  }
}

# HERE macro is deprecated
# auto-remove common cases
/HERE/ {
  gsub("<< HERE \"", "\"");
  gsub(", HERE\)", ", MYNAME\)");
}
```

With rules added as we agree on changes (discussion in the PR adding the rule), and removed when we think it no longer needed (eg HERE actually deleted, so code containing it wont even build)

Reply:
If your plans materialize, we will have at least two scripts, one (existing) formatting sources and one (future) removing deprecated stuff. In that case, the iteration I suggested would iterate over those two scripts. We can hard-code that iteration (essentially like this PR is currently doing, but polished to address my other concerns):

```
apply maintenance/remove-deprecated.awk
apply maintenance/format-sources.pl
```

or generalize it (naming each of the two scripts to match some suitable shell pattern):

```
for script in maintenance/cpp-*
do
    apply $script
done
```

There are (obvious?) pros and cons for each approach. I suggested the latter because it makes adding dummy scripts unnecessary and makes this PR useful for generalizing safe script execution. I can live with hard-coding if others prefer or you insist on that.

I hope that this clarifies my change request.

Reply:
>  I can live with hard-coding if others prefer or you insist on that.

... but for this "provide infrastructure for future scripts" PR to be useful, even with hard-coding, it most likely should still follow the (first) `apply ...` approach sketched above, even if it will apply just one (existing) script. With that approach, there will be no reason to add a dummy script execution -- it would be trivial to add one more `apply ...` line later if/as needed.